### ISSUE TITLE
Question: MCMC being "sticky" or stuck across different seeds

### ISSUE THREAD

**Issue by leetim13** [ocbe-uio/BayesMallows#307]
Not sure if this is the best place to post this, but we are running into some scenarios where the posterior distribution is having inconsistent results across different seeds (see attached alpha and rho plots). Looking for any potential recommendations/suggestions!

The data has `71 n_items` (i.e., "companies") with pair-wise rankings and was fitted using both `kendall` and `footrule` distances which produced similar results (we ended up picking kendall to avoid calculating the partition function). We also tried to fine-tune the bmm model using the **following hyper-parameters**, but the stickiness results are still the same (e.g., for company "Genomic Health Inc." in the third row, the posterior is centered at either very left or very right depending on the seed number across different parameter settings).

```
bmm_nmc_values <- c(100000) 
leap_size_values <- c(floor(n_items / 2), floor(n_items / 5), floor(n_items / 7))
alpha_prop_sd_values <- c(0.1, 0.5)
rho_thinning_values <- c(100)
alpha_jump_values <- c(10)
lambda_values <- c(0.001, 0.01, 0.1, 0.2)
bmm_burnin <- bmm_nmc*0.5 
```

We also tried switching `shuffle_unranked` and `random` in `generate_initial_ranking`, but `random=TRUE` takes too long given `71 n_items` and not sure if this could be the root cause. Judging by the alpha trace plot, it also seems like the MCMC has somewhat converged and increasing the number of iterations to `1 million` didn't help as well. Any feedback/suggestions would be greatly appreciated!

![S0221_2016_posteriordensity_rho_v_burn_5e+050_seed103_20230903](https://github.com/ocbe-uio/BayesMallows/assets/20623399/5f073317-d20a-4a2e-9250-d75be9d4ddea)
![S0221_2016_posteriordensity_rho_v_burn_5e+050_seed102_20230903](https://github.com/ocbe-uio/BayesMallows/assets/20623399/ee2e7733-5612-47cc-881f-e35e90152d17)
![S0221_2016_posteriordensity_rho_v_burn_5e+050_seed101_20230903](https://github.com/ocbe-uio/BayesMallows/assets/20623399/52c3a7fa-5871-48c9-b9a8-a734894e1f7e)

![S0221_2016_trace_alpha_v0_seed103_20230903](https://github.com/ocbe-uio/BayesMallows/assets/20623399/694a69ed-74f9-4e90-8d1a-82320f782283)
![S0221_2016_trace_alpha_v0_seed102_20230903](https://github.com/ocbe-uio/BayesMallows/assets/20623399/b0f2f4b1-f1be-40d5-888b-5811f505d85f)
![S0221_2016_trace_alpha_v0_seed101_20230903](https://github.com/ocbe-uio/BayesMallows/assets/20623399/e0b071f3-2093-41c3-85bb-d8e4a6e8c2b6)

**Comment by osorensen** [comment]
Thanks for reaching out! This is good place to post the question.

Your alpha plots seem consistent across runs. This seems to me to suggest that with different seeds, you end up getting stuck in different configurations that have very similar "goodness of fit". The problem of course is that if MCMC converged properly, it should visit both (or all) of these configurations. Instead it gets stuck. Considering the example of "Genomic Health Inc.", it seems like the transition from one state to the other involves some MCMC steps with very low probability. Given what you see I wonder if running, say, 10 parallel MCMC chains would be a good and pragmatic choice here, and then simply combine all the posterior samples after burn-in.

I'll ask some of my colleagues who are more experienced in working with these models for further help. They'll get back to you in this thread.

**Comment by osorensen** [comment]
@leetim13, having discussed the issue a bit with the colleague, I think running multiple MCMC chains is the best idea at the moment. However, I realize that this currently is not an option in the `compute_mallows` function, and hence would require some low-level coding on your part in order to visualize the posterior distributions. Luckily, this should be relatively easy to add, so I'll open another issue and hopefully we have it fixed within a week or two.

**Comment by leetim13** [comment]
@osorensen, thanks for the quick response! Just to clarify, for now, this can be achieved by using `set.seed()` to represent the multiple chains, then concatenating all posterior distributions for visualization purposes?

**Comment by osorensen** [comment]
Indeed @leetim13. You don't necessarily need to set the seed, but if you set it, it should be different for each run.

Below is an example showing you how to do it with the current version of BayesMallows. I run the chains in parallel, and then but the combined output into a dummy model so that the plotting functions still work.

``` r
library(BayesMallows)

beach_tc <- generate_transitive_closure(beach_preferences)

# I run three parallel chains and save in a list.
model_fits <- replicate(3, {
  # Should generate this one randomly each time
  beach_rankings <- generate_initial_ranking(beach_tc, n_items = 15,
                                             shuffle_unranked = TRUE)
  
  model_fit <- compute_mallows(rankings = beach_rankings,
                               preferences = beach_tc,
                               save_aug = TRUE)
  
  model_fit$burnin <- 500
  model_fit
}, simplify = FALSE)

# You can plot each model fit separately, but should be able to combine them.
plot(model_fits[[1]], parameter = "rho")
#> Items not provided by user. Picking 5 at random.
```

![](https://i.imgur.com/3LhKkgd.png)

``` r

# Here I combine the posterior samples
posterior_rho <- do.call(rbind, lapply(model_fits, function(x){
  x$rho[x$rho$iteration < x$burnin, ]
}))

head(posterior_rho)
#>     item   cluster iteration value
#> 1 Item 1 Cluster 1         1     2
#> 2 Item 2 Cluster 1         1    15
#> 3 Item 3 Cluster 1         1     9
#> 4 Item 4 Cluster 1         1    11
#> 5 Item 5 Cluster 1         1    13
#> 6 Item 6 Cluster 1         1     5

posterior_alpha <- do.call(rbind, lapply(model_fits, function(x){
  x$alpha[x$alpha$iteration < x$burnin, ]
}))

head(posterior_alpha)
#>     cluster iteration     value
#> 1 Cluster 1         1 1.0000000
#> 2 Cluster 1         2 1.0348185
#> 3 Cluster 1         3 0.8925915
#> 4 Cluster 1         4 0.8453198
#> 5 Cluster 1         5 0.9580004
#> 6 Cluster 1         6 0.9611377

# Now I fit a fake model
beach_rankings <- generate_initial_ranking(beach_tc, n_items = 15,
                                           shuffle_unranked = TRUE)
dummy_model <- compute_mallows(rankings = beach_rankings,
                               preferences = beach_tc,
                               save_aug = TRUE, nmc = 20)

dummy_model$rho <- posterior_rho
dummy_model$alpha <- posterior_alpha
dummy_model$burnin <- 0

# Now I can plot the full posteriors based on all models in model_fits
plot(dummy_model, parameter = "rho")
#> Items not provided by user. Picking 5 at random.
```

![](https://i.imgur.com/1IJYo3k.png)

``` r
plot(dummy_model, parameter = "alpha")
```

![](https://i.imgur.com/WU1iGdo.png)

<sup>Created on 2023-09-07 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by osorensen** [comment]
👋 @leetim13, this fix is now on the master branch, and will be on CRAN within the next weeks.
